Handle extension resources on server

sandy081 opened the pull request:
Handle extension resources on server

alexdima commented on src/vs/server/node/webClientServer.ts:
```diff
@@ -100,6 +107,10 @@ export class WebClientServer {
 				// callback support
 				return this._handleCallback(res);
 			}
+			if (/^\/extensionResource\//.test(pathname)) {
```
Should we rename this to `web-extension-resource`? Lowercase to align with typical paths used in HTTP requests, and prepending `web-` to clarify that this is only about web worker extensions i.e. not about extensions installed on the server, which might also have resources.

sandy081 replied:
Done

alexdima commented on src/vs/server/node/webClientServer.ts:
```diff
@@ -128,6 +139,38 @@ export class WebClientServer {
+
+		const url = URI.parse(path).with({
+			scheme: this._productService.extensionsGallery?.resourceUrlTemplate ? URI.parse(this._productService.extensionsGallery.resourceUrlTemplate).scheme : 'https',
+			authority: path.substring(0, path.indexOf('/')),
```
Could the authority be hard-coded to the authority of `productService.extensionsGallery.resourceUrlTemplate`? If I understand this correctly, the current implementation would transform `/extensionResource/example.com` to `example.com` and it might be better to avoid that.

sandy081 replied:
It can be hardcoded because the authority is scoped to publisher. For eg it looks like follows:

`https://vscodevim.vscode-unpkg.net/vscodevim/vim/1.21.10/extension/out/extensionWeb.js`

monken replied:
are you protecting for private hostnames and IPs? This could open up Server Side Request Forgery (i.e. ability to retrieve credentials from a cloud vendors metadata service).

sandy081 replied:
Even though this redirection is for a specific service, added a check to make sure the redirection url is pointing to the expected service

alexdima reviewed (approved):
Nice! I just left a couple of suggestions.

monken commented on src/vs/server/node/webClientServer.ts:
```diff
@@ -128,6 +139,38 @@ export class WebClientServer {
+
+		res.writeHead(context.res.statusCode || 500, context.res.headers);
+		const buffer = await streamToBuffer(context.stream);
+		return res.end(buffer.buffer);
```
use `context.stream.pipe(res)` instead to reduce memory pressure on the server? `streamToBuffer` implementation will read the whole file into memory which should be avoided.

bpasero replied:
👍

sandy081 replied:
`context.stream` is our custom `VSBufferReadableStream` and is not supporting `pipe`.

monken commented on src/vs/server/node/webClientServer.ts:
```diff
@@ -128,6 +139,38 @@ export class WebClientServer {
+		const headers: IHeaders = {};
+		for (const header of req.rawHeaders) {
+			if (req.headers[header]) {
+				headers[header] = req.headers[header]![0];
```
copying all headers seems risky. This will include potentially sensitive headers such as `Cookie`.

bpasero replied:
😱

sandy081 replied:
Changed it to send selective headers

monken commented:
What’s the rationale for this feature? It seems to create a completely open reverse proxy to any domain on the internet (and worse, the private networks that the server has access to). The browser handles cross origin request in a separate security context (which is what you want), this feature will effectively disable these safeguards.

bpasero commented on src/vs/server/node/webClientServer.ts:
```diff
@@ -183,12 +226,23 @@ export class WebClientServer {
+					extensionsGallery: resourceUrlTemplate ? {
+						...this._productService.extensionsGallery,
+						'resourceUrlTemplate': resourceUrlTemplate.with({
+							scheme: 'http',
```
What if the server runs with `https`, isn't this dynamic?

sandy081 replied:
True, since server cannot detect the protocol, this is handled on the client side.

bpasero commented on src/vs/server/node/webClientServer.ts:
```diff
@@ -183,12 +226,23 @@ export class WebClientServer {
 			accessToken: this._environmentService.args['github-auth'],
 			scopes: [['user:email'], ['repo']]
 		} : undefined;
+		const resourceUrlTemplate = this._productService.extensionsGallery ? URI.parse(this._productService.extensionsGallery.resourceUrlTemplate) : undefined;
```
Is the idea of `resourceUrlTemplate` to be a prefix that paths are being concatenated to?

sandy081 replied:
Yes, for eg:

https://github.com/microsoft/vscode/blob/83d42f8877a72133dc76b151195a98d21ab36717/src/vs/workbench/services/extensionManagement/browser/webExtensionsScannerService.ts#L401

bpasero commented on src/vs/server/node/webClientServer.ts:
```diff
@@ -128,6 +139,38 @@ export class WebClientServer {
+	 */
+	private async _handleWebExtensionResource(req: http.IncomingMessage, res: http.ServerResponse, parsedUrl: url.UrlWithParsedQuery): Promise<void> {
+		// Strip `/web-extension-resource/` from the path
+		const normalizedPathname = decodeURIComponent(parsedUrl.pathname!); // support paths that are uri-encoded (e.g. spaces => %20)
```
This is so involved to get back to the original URL, can this be encoded in a query parameter rather? Or is the template a strict prefix that paths are concatenated to.

sandy081 replied:
> Or is the template a strict prefix that paths are concatenated to.

Yes. Hence it is made part of the `path` like `static` urls.

bpasero commented on src/vs/server/node/webClientServer.ts:
```diff
@@ -128,6 +139,38 @@ export class WebClientServer {
+		const path = normalize(normalizedPathname.substr('/web-extension-resource/'.length));
+
+		const url = URI.parse(path).with({
+			scheme: this._productService.extensionsGallery?.resourceUrlTemplate ? URI.parse(this._productService.extensionsGallery.resourceUrlTemplate).scheme : 'https',
```
`URI.parse(this._productService.extensionsGallery.resourceUrlTemplate).scheme` can probably be cached?

sandy081 commented:
@monken Thanks for your feedback and very much appreciated. This is not targeting any server/domain in the internet. This is to redirect requests to the service (we are hosting) from the client to the server. We have strict CORS rules on this service of ours and its blocking the dev set up and testing. Instead of loosening the CORS rules (which is not ideal) we are redirecting these specific requests through server. Hope it is clear.

sandy081 commented:
Updated the PR to support following:

- Added a check on the server to validate if the URL is pointing to the target service or not.
- Selectively set request and response headers
- Client updates the protocol (`http` or `https`) of the URL correctly.

@alexdima @bpasero Please take a look at the changes.
